examples/shipping/booking/instrumenting.go

```diff
@@ -0,0 +1,114 @@
+package booking
+
+import (
+	"time"
+
+	stdprometheus "github.com/prometheus/client_golang/prometheus"
+
+	"github.com/go-kit/kit/metrics"
+	kitprometheus "github.com/go-kit/kit/metrics/prometheus"
+
+	"github.com/go-kit/kit/examples/shipping/cargo"
+	"github.com/go-kit/kit/examples/shipping/location"
+)
+
+type instrumentingService struct {
+	requestCount   metrics.Counter
+	requestLatency metrics.TimeHistogram
+	Service
+}
+
+// NewInstrumentingService returns an instance of an instrumenting Service.
+func NewInstrumentingService(s Service) Service {
+	fieldKeys := []string{"method"}
+
+	requestCount := kitprometheus.NewCounter(stdprometheus.CounterOpts{
+		Namespace: "api",
+		Subsystem: "booking_service",
+		Name:      "request_count",
+		Help:      "Number of requests received.",
+	}, fieldKeys)
+
+	requestLatency := metrics.NewTimeHistogram(time.Microsecond, kitprometheus.NewSummary(stdprometheus.SummaryOpts{
+		Namespace: "api",
+		Subsystem: "booking_service",
+		Name:      "request_latency_microseconds",
+		Help:      "Total duration of requests in microseconds.",
+	}, fieldKeys))
+
+	return &instrumentingService{
+		requestCount:   requestCount,
+		requestLatency: requestLatency,
+		Service:        s,
+	}
+}
```

Comment:
Hmm. I tend to take the metrics as dependencies, rather than constructing them in the middleware, as you sometimes need to manage the registration (via the Prometheus HTTP handler) in unique ways. But I reckon that's a matter of opinion.